### PR TITLE
graph/topo: add undirected cycles routine

### PR DESCRIPTION
**Pull request by kortschak**
@mewmew @vladimir-ch Please take a look.

Depends on #70.

Fixes gonum/graph#197.

### CODE REVIEW

**Comment by mewmew** [comment]
Glad to see you are implementing these kind of algorithms. Truthfully, I'm still in the learning process of graph theory, and feel I have yet some way to go before being able to review this with confidence. Thus leaving this one for @vladimir-ch.

This autumn I'll attend a course in graph theory at Uni, which should hopefully put me in a better position for the future regarding these kind of algorithms.

**Comment by kortschak** [comment]
@vladimir-ch *ping* when you have time.

File: graph/topo/paton_cycles.go [hunk 1]
```diff
@@ -0,0 +1,86 @@
...
+
+package topo
+
+import (
+	"gonum.org/v1/gonum/graph"
+	"gonum.org/v1/gonum/graph/internal/linear"
+	"gonum.org/v1/gonum/graph/internal/set"
+)
+
+// UndirectedCyclesIn returns the set of cycle bases in the graph g.
```
Comment (vladimir-ch) [inline graph/topo/paton_cycles.go]: For me the basis would be the set itself, and the cycles would then be called basis cycles. I would maybe change the comment to
```
UndirectedCyclesIn returns a set of cycles that forms a cycle basis in the graph g.
Any cycle in g can be constructed as a symmetric difference of its elements.
```

File: graph/topo/paton_cycles_test.go [hunk 2]
```diff
@@ -0,0 +1,131 @@
...
+	"gonum.org/v1/gonum/graph/internal/ordered"
+	"gonum.org/v1/gonum/graph/simple"
+)
+
+var undirectedCyclesInTests = []struct {
+	g    []intset
+	want [][][]int64
+}{
+	{
+		g: []intset{
```
Comment (vladimir-ch) [inline graph/topo/paton_cycles_test.go]: Would it be worth adding to this graph another component that has a cycle? Like:
```
10: linksTo(11, 12),
11: linksTo(12),
12: nil,
```

File: graph/topo/paton_cycles_test.go [hunk 3]
```diff
@@ -0,0 +1,131 @@
...
+				break
+			}
+		}
+		if !matched {
+			t.Errorf("unexpected paton result for %d:\n\tgot:%#v\n\twant from:%#v", i, got, test.want)
+		}
+	}
+}
+
+func canonicalise(c []graph.Node) []graph.Node {
```
Comment (kortschak) [inline graph/topo/paton_cycles_test.go]: Done.

File: graph/topo/paton_cycles.go [hunk 4]
```diff
@@ -0,0 +1,86 @@
...
+
+		for tree.Len() != 0 {
+			u := tree.Pop()
+			uid := u.ID()
+			adj := from[uid]
+			for _, v := range g.From(u) {
+				vid := v.ID()
+				switch {
+				case uid == vid:
+					panic("topo: found self loop")
```
Comment (kortschak) [inline graph/topo/paton_cycles.go]: This is removed since we now allow user graphs to contain self-loops.

File: graph/topo/paton_cycles.go (hunk 1 above)
Comment (kortschak) [inline graph/topo/paton_cycles.go]: Done.

File: graph/topo/paton_cycles_test.go (hunk 2 above)
Comment (kortschak) [inline graph/topo/paton_cycles_test.go]: Done.

**Comment by kortschak** [comment]
PTAL to re-approve.

**Comment by kortschak** [comment]
Probably I should add a `Xor(cycles ...[]graph.Node) [][]graph.Node`/`SymDiff(cycles ...[]graph.Node) [][]graph.Node` function to allow people to reconstruct any given set of cycles. This would be a new PR.

**Pull request merged**